### PR TITLE
added(server): api endpoint for replacing batch of user topics with new one.

### PR DESCRIPTION
**Pull request by mmoehabb**

### CODE REVIEW

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-8t7yt8zgq-litespace-core.vercel.app) 
🚀 [Luna](https://luna-emswkf77m-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-k8c3w44zu-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-5a4il9ppe-litespace-core.vercel.app)

File: services/server/src/handlers/topic.ts [hunk 1]
```diff
@@ -31,6 +32,12 @@ const generalUserTopicsBodyParser = zod.object({
   topicIds: zod.array(zod.number()),
 });
 
+const replaceUserTopicsBodyParser = zod.object({
+  oldTopics: zod.array(zod.number()),
+  newTopics: zod.array(zod.number()),
```
Comment (neuodev) [inline services/server/src/handlers/topic.ts:37]: You can accept just the new topics. You can get all the old topics from the database. 

```suggestion
  topics: zod.array(id),
```

**Review by neuodev — CHANGES_REQUESTED** [review]

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-6ryi179pl-litespace-core.vercel.app) 
🚀 [Luna](https://luna-836qau3rq-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-c3vzkp6i8-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-3wiokal2h-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-ehfw1cyth-litespace-core.vercel.app) 
🚀 [Luna](https://luna-ig5p5llg1-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-hhkm1puu5-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-78sub0czj-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-l5rlnvvvj-litespace-core.vercel.app) 
🚀 [Luna](https://luna-8ny7sk1py-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-5mpfvph7u-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-azvxd64n6-litespace-core.vercel.app)

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-f0ryzzpuh-litespace-core.vercel.app) 
🚀 [Luna](https://luna-aan49kde2-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-96ck7g8av-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-51819tl61-litespace-core.vercel.app)

File: services/server/src/handlers/topic.ts [hunk 2]
```diff
@@ -170,6 +178,46 @@ async function addUserTopics(req: Request, res: Response, next: NextFunction) {
...
+  const allowed = isStudent(user) || isTutor(user) || isTutorManager(user);
+  if (!allowed) return next(forbidden());
+
+  const { removeTopics, addTopics }: ITopic.ReplaceUserTopicsApiPayload = 
+    replaceUserTopicsBodyParser.parse(req.body);
+
+  if (isEmpty(removeTopics) || isEmpty(addTopics)) return next(empty());
+  // to not leave a vulnerability which can make the user exceed the max number
+  if (removeTopics.length !== addTopics.length) return next(bad()); 
+
```
Comment (neuodev) [inline services/server/src/handlers/topic.ts:192]: - It is fine to one of them to be empty. User might want to just add or remove but not both at the same time.
- No need for the second check (I don't understand it)

```suggestion
  if (isEmpty(removeTopics) && isEmpty(addTopics)) return next(empty());
```

File: services/server/src/handlers/topic.ts [hunk 3]
```diff
@@ -170,6 +178,46 @@ async function addUserTopics(req: Request, res: Response, next: NextFunction) {
   res.sendStatus(200);
 }
 
+async function replaceUserTopics(req: Request, res: Response, next: NextFunction) {
+  const user = req.user;
+  const allowed = isStudent(user) || isTutor(user) || isTutorManager(user);
+  if (!allowed) return next(forbidden());
+
+  const { removeTopics, addTopics }: ITopic.ReplaceUserTopicsApiPayload = 
+    replaceUserTopicsBodyParser.parse(req.body);
```
Comment (neuodev) [inline services/server/src/handlers/topic.ts:187]: for consistency


```suggestion
  const { removeTopics, addTopics }: ITopic.ReplaceUserTopicsApiPayload = 
    replaceUserTopicsPayload.parse(req.body);
```

File: services/server/src/handlers/topic.ts [hunk 4]
```diff
@@ -170,6 +178,46 @@ async function addUserTopics(req: Request, res: Response, next: NextFunction) {
...
+
+  if (isEmpty(removeTopics) || isEmpty(addTopics)) return next(empty());
+  // to not leave a vulnerability which can make the user exceed the max number
+  if (removeTopics.length !== addTopics.length) return next(bad()); 
+
+  // ensure that all oldTopics exists for the user
```
Comment (neuodev) [inline services/server/src/handlers/topic.ts:193]: You can verify that the user didn't exceeded the max allowed topic count by following this simple equation.

```
const allowed = (inDbCount + addTopics.legnth - removeTopics.legnth) <= MAX_ALLOWED_TOPICS 
```

**Review by neuodev — CHANGES_REQUESTED** [review]

File: services/server/src/handlers/topic.ts (hunk 2 above)
Comment (mmoehabb) [reply services/server/src/handlers/topic.ts:192]: I consider just adding or removing topics (by the user) should be handled by other functions (each corresponding handler). using this handler generically is unnecessarily confusing and may be harmful as well. The frontend should use specific and different handlers (endpoints) for different kinds of manipulation: add, remove, update, and replace.

Comment (neuodev) [reply services/server/src/handlers/topic.ts:192]: Due to how the frontend currently works, this feature is required to work this way. Eventually we will remove the other routes and keep just this new handler.

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-ckdytt02v-litespace-core.vercel.app) 
🚀 [Luna](https://luna-pbxoccglt-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-1rak3vn94-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-i41r8a4sk-litespace-core.vercel.app)

File: services/server/src/handlers/topic.ts [hunk 5]
```diff
@@ -170,12 +178,54 @@ async function addUserTopics(req: Request, res: Response, next: NextFunction) {
...
+  const user = req.user;
+  const allowed = isStudent(user) || isTutor(user) || isTutorManager(user);
+  if (!allowed) return next(forbidden());
+
+  const { removeTopics, addTopics }: ITopic.ReplaceUserTopicsApiPayload = 
+    replaceUserTopicsPayload.parse(req.body);
+
+  if (isEmpty(removeTopics) && isEmpty(addTopics)) return next(empty());
+
+  // ensure that all oldTopics exists for the user
```
Comment (neuodev) [inline services/server/src/handlers/topic.ts:191]:
```suggestion
  // ensure that all topics in `addTopics` exists for the current user
```

File: services/server/src/handlers/topic.ts [hunk 6]
```diff
@@ -170,12 +178,54 @@ async function addUserTopics(req: Request, res: Response, next: NextFunction) {
...
+  }
+  
+  // verify that all topics do exist in the database
+  const isExists = await topics.isExistsBatch(addTopics);
+  if (Object.values(isExists).includes(false))
+    return next(notfound.topic());
+
+  // ensure that user topics will not exceed the max num after insertion
+  if ((inDBTopicIds.length + addTopics.length - removeTopics.length) > MAX_TOPICS_COUNT)
+    return next(refused());
```
Comment (neuodev) [inline services/server/src/handlers/topic.ts:206]: The code will be cleaner if we extracted this logic into a separate variable.

 ```suggestion
  // ensure that user topics will not exceed the max num after insertion
  const exceededMaxAllowedTopics = (inDBTopicIds.length + addTopics.length - removeTopics.length) > MAX_TOPICS_COUNT
  if (exceededMaxAllowedTopics) return next(refused());
```

File: services/server/src/handlers/topic.ts [hunk 7]
```diff
@@ -170,12 +178,54 @@ async function addUserTopics(req: Request, res: Response, next: NextFunction) {
...
+  await knex.transaction(async (tx: Knex.Transaction) => {
+    isEmpty(removeTopics) || await topics.deleteUserTopics({
+      user: user.id,
+      topics: removeTopics,
+      tx,
+    });
+    isEmpty(addTopics) || await topics.registerUserTopics({
+      user: user.id,
+      topics: addTopics
+    }, tx);
```
Comment (neuodev) [inline services/server/src/handlers/topic.ts:217]: Cleaner.

```suggestion
    if (!isEmpty(removeTopics)) await topics.deleteUserTopics({
      user: user.id,
      topics: removeTopics,
      tx,
    });
    
    if(!isEmpty(addTopics)) await topics.registerUserTopics({
      user: user.id,
      topics: addTopics
    }, tx);
```

**Review by neuodev — CHANGES_REQUESTED** [review]

**Comment by neuodev** [comment]
🚀 [Nova](https://nova-mnplaa68n-litespace-core.vercel.app) 
🚀 [Luna](https://luna-d3u05tfng-litespace-core.vercel.app) 
🚀 [Nebula](https://nebula-b24rlw1cq-litespace-core.vercel.app) 
🚀 [Dashboard](https://dashboard-qiwbe79jz-litespace-core.vercel.app)

**Review by neuodev — APPROVED** [review]

**Pull request merged**